File: src/main/java/com/jore/epoc/bo/orders/BuildFactoryOrder.java
package com.jore.epoc.bo.orders;

import org.hibernate.annotations.CompositeType;

import com.jore.datatypes.money.Money;
import com.jore.epoc.bo.Factory;
import com.jore.epoc.bo.accounting.FinancialAccounting;
import com.jore.epoc.bo.message.MessageLevel;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;

@Entity
public class BuildFactoryOrder extends AbstractSimulationOrder {
    private Integer productionLines;
    private Integer timeToBuild;
    private Integer dailyCapacityPerProductionLine;
    @AttributeOverride(name = "amount", column = @Column(name = "labor_cost_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "labor_cost_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money productionLineLaborCost;
    @AttributeOverride(name = "amount", column = @Column(name = "fixed_cost_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "fixed_cost_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money constructionCosts;
    @AttributeOverride(name = "amount", column = @Column(name = "variable_cost_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "variable_cost_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money constructionCostsPerLine;

    @Override
    public void execute() {
        Money factoryCosts = constructionCosts.add(constructionCostsPerLine.multiply(productionLines));
        if (getCompany().getAccounting().checkFunds(factoryCosts, getExecutionMonth().atEndOfMonth())) {
            addFactory();
            book(getExecutionMonth().atDay(FIRST_OF_MONTH), "Built factory", FinancialAccounting.REAL_ESTATE, FinancialAccounting.BANK, constructionCosts.add(constructionCostsPerLine.multiply(productionLines)));
            addMessage(MessageLevel.INFORMATION, "FactoryCreated", factoryCosts, getExecutionMonth());
            setExecuted(true);
        } else {
            addMessage(MessageLevel.WARNING, "NoFactoryDueToFunds", getExecutionMonth(), factoryCosts, getCompany().getAccounting().getBankBalance(getExecutionMonth().atEndOfMonth()));
        }
    }

    @Override
    public int getSortOrder() {
        return 4;
    }

    public void setConstructionCost(Money constructionCosts) {
        this.constructionCosts = constructionCosts;
    }

    public void setConstructionCostPerLine(Money constructionCostsPerLine) {
        this.constructionCostsPerLine = constructionCostsPerLine;
    }

    public void setDailyCapacityPerProductionLine(Integer dailyCapacityPerProductionLine) {
        this.dailyCapacityPerProductionLine = dailyCapacityPerProductionLine;
    }

    public void setProductionLineLaborCost(Money productionLineLaborCost) {
        this.productionLineLaborCost = productionLineLaborCost;
    }

    public void setProductionLines(int productionLines) {
        this.productionLines = productionLines;
    }

    public void setTimeToBuild(Integer timeToBuild) {
        this.timeToBuild = timeToBuild;
    }

    private void addFactory() {
        Factory factory = new Factory();
        factory.setProductionLines(productionLines);
        factory.setProductionStartMonth(getExecutionMonth().plusMonths(timeToBuild));
        factory.setDailyCapacityPerProductionLine(dailyCapacityPerProductionLine);
        factory.setProductionLineLaborCost(productionLineLaborCost);
        getCompany().addFactory(factory);
    }
}

File: EpocServer/Sources/App/Models/Orders/BuildFactoryOrder.swift
import Foundation

public final class BuildFactoryOrder: SimulationOrder {
    public var id: Int64?
    public var executionMonth: YearMonth = .of(2020, 1)
    public var isExecuted: Bool = false
    public weak var company: Company!
    public var productionLines: Int = 10
    public var timeToBuild: Int = 0
    public var dailyCapacityPerProductionLine: Int = 4
    public var productionLineLaborCost: Money = Money.of("CHF", 500000)
    public var constructionCost: Money = Money.of("CHF", 1000000)
    public var constructionCostPerLine: Money = Money.of("CHF", 100000)

    public var sortOrder: Int { 4 }
    public var type: String { "Build Factory" }

    public func getAmount() -> Money {
        constructionCost.add(constructionCostPerLine.multiply(productionLines))
    }

    public func execute() {
        if company.accounting.checkFunds(getAmount(), valueDate: executionMonth.atEndOfMonth()) {
            addFactory()
            book(bookingDate: executionMonth.atDay(1), bookingText: "Built factory",
                 debitAccount: FinancialAccounting.REAL_ESTATE, creditAccount: FinancialAccounting.BANK,
                 amount: constructionCost.add(constructionCostPerLine.multiply(productionLines)))
            addMessage(.information, "FactoryCreated", getAmount(), executionMonth)
            isExecuted = true
        } else {
            addMessage(.warning, "NoFactoryDueToFunds", executionMonth, getAmount(), company.accounting.getBankBalance(executionMonth.atEndOfMonth()))
        }
    }

    private func addFactory() {
        let factory = Factory()
        factory.productionLines = productionLines
        factory.productionStartMonth = executionMonth.plusMonths(timeToBuild)
        factory.dailyCapacityPerProductionLine = dailyCapacityPerProductionLine
        factory.productionLineLaborCost = productionLineLaborCost
        company.addFactory(factory)
    }
}
